src/Http/Headers/src/ContentDispositionHeaderValue.cs

```diff
@@ -625,48 +626,55 @@ private static string Encode5987(StringSegment input)
             ? stackalloc byte[MaxStackAllocSizeBytes]
             : bufferFromPool = ArrayPool<byte>.Shared.Rent(maxInputBytes);
 
+        var maxInputAsciiCount = Encoding.ASCII.GetMaxCharCount(input.Length);
+        char[]? charBufferFromPool = null;
+        Span<char> tempCharBuffer = maxInputAsciiCount <= MaxStackAllocSizeChars
+            ? stackalloc char[MaxStackAllocSizeChars]
+            : charBufferFromPool = ArrayPool<char>.Shared.Rent(maxInputAsciiCount);
+
         var bytesWritten = Encoding.UTF8.GetBytes(input, inputBytes);
         inputBytes = inputBytes[..bytesWritten];
 
-        int totalBytesConsumed = 0;
-        while (totalBytesConsumed < inputBytes.Length)
+        while (inputBytes.Length > 0)
         {
-            if (Ascii.IsValid(inputBytes[totalBytesConsumed]))
+            var length = inputBytes.IndexOfAnyExcept(Rfc5987AttrChar);
+            if (length < 0)
             {
-                // This is an ASCII char. Let's handle it ourselves.
-
-                char c = (char)inputBytes[totalBytesConsumed];
-                if (!AttrChar.Contains(c))
-                {
-                    HexEscape(builder, c);
-                }
-                else
-                {
-                    builder.Append(c);
-                }
-
-                totalBytesConsumed++;
+                length = inputBytes.Length;
             }
-            else
-            {
-                // Non-ASCII, let's rely on Rune to decode it.
+            int asciiCharCount = Encoding.ASCII.GetChars(inputBytes.Slice(0, length), tempCharBuffer);
```

Comment:
Could we avoid the pool rental by chunking `inputBytes` so that the individual chunks fit within the stackalloc'd buffer?

Reply:
I will do a comparison

Reply:
I tried avoiding the string -> UTF8 -> ASCII encoding in the case we know it is a limited set of ASCII chars only, but apparently that does not make it faster. I think it could be related to `SearchValues<>` closed generic on `char` vs. `byte`.

I am still following up on the commented suggestion.

Reply:
So if I understod what you suggest the `tempCharBuffer` allocation changes to:

```csharp
Span<char> tempCharBuffer = stackalloc char[input.Length <= MaxStackAllocSizeChars ? input.Length : MaxStackAllocSizeChars];
```
and the Ascii encoding:

```suggestion
 var toAscii = inputBytes.Slice(0, length);
 while (toAscii.Length > 0)
 {
     var chunkSize = toAscii.Length <= tempCharBuffer.Length ? toAscii.Length : tempCharBuffer.Length;
     int asciiCharCount = Encoding.ASCII.GetChars(toAscii.Slice(0, chunkSize), tempCharBuffer);
     builder.Append(tempCharBuffer[..asciiCharCount]);
     toAscii = toAscii.Slice(chunkSize);
 }
```
My performance results:

|                 Method |     Mean |   Error |  StdDev |        Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |---------:|--------:|--------:|------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 235.9 ns | 2.38 ns | 1.99 ns | 4,239,131.0 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding | 109.2 ns | 0.53 ns | 0.44 ns | 9,156,655.4 | 0.0039 |     - |     - |     368 B |

Reply:
No, it would be something like
```
var asciiBuffer = Unsafe.As<byte, char>(inputBytes.Slice(0, length));
builder.Append(asciiBuffer);
```

That code is likely wrong, but it's the general idea.

Reply:
I am thinking about `MemoryMarshal.Cast<byte, char>(inputBytes.Slice(0, length));`, but I wonder if this will be correct, because UTF-8 bytes uses 1 byte for a character and `char` 2 bytes. So, reinterpreting the byte array would not be correct, would it?

Reply:
It's certainly not intuitively correct, but it's not impossible that there's special handling for this case.

Reply:
`MemoryMarshal.Cast<byte, char>(inputBytes.Slice(0, length));` got me gibberish, I am looking for alternatives still.

Reply:
I am looking at an alternative approach promises some benefits, I will polish the code and push it over the weekend

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 224.93 ns | 1.833 ns | 1.715 ns |  4,445,824.9 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  87.70 ns | 0.887 ns | 0.787 ns | 11,402,554.9 | 0.0039 |     - |     - |     368 B |

Reply:
### Step 1 - commit [Using Rune to encode/decode non ascii characters](https://github.com/dotnet/aspnetcore/pull/55039/commits/7a7f2fd20d64ef551d03fd9ea2d3497783828151)

I have implemented the idea I suggested above. See commit  where the idea is to 
1. For Rfc5987AttrChar append characters to `StringBuilder` directly
1. For non-Rfc5987AttrChars keep using Rune to encode single character-by-character (or 2-chars by 2-chars (Rune)).

The results are:

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 221.70 ns | 1.567 ns | 1.389 ns |  4,510,656.9 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  89.11 ns | 0.905 ns | 0.802 ns | 11,221,787.5 | 0.0039 |     - |     - |     368 B |


### Step 2 - commit [With custom utf8 and hex encoding combined](https://github.com/dotnet/aspnetcore/pull/55039/commits/2eadb693477f2556fc50787575bc325b3b22c6de)

Step 2 takes this a step forward and "borrows" the implementation of [Rune's EncodeToUtf8](https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Text/Rune.cs,6487d779e626351e,references) and combines it with the hex creation. This avoids the inner loop around the Hex creation.

My best measured results using this approach:

|                 Method |      Mean |    Error |   StdDev |         Op/s |  Gen 0 | Gen 1 | Gen 2 | Allocated |
|----------------------- |----------:|---------:|---------:|-------------:|-------:|------:|------:|----------:|
|   FileNameStarEncoding | 213.61 ns | 1.828 ns | 1.710 ns |  4,681,410.7 | 0.0052 |     - |     - |     496 B |
| FileNameStarNoEncoding |  87.12 ns | 0.962 ns | 0.853 ns | 11,478,216.6 | 0.0039 |     - |     - |     368 B |